platform/cog-platform-drm.c

```diff
@@ -236,6 +279,180 @@ init_egl ()
+        event.id = input_data.last_touch_id;
+        event.time = time;
+        event.modifiers = 0;
+
+        wpe_view_backend_dispatch_touch_event (wpe_view_data.backend, &event);
+
+        for (int i = 0; i < 10; ++i) {
+            struct wpe_input_touch_event_raw *touch_point = &input_data.touch_points[i];
+            if (touch_point->type != wpe_input_touch_event_type_up)
+                continue;
+
+            memset (touch_point, 0, sizeof (struct wpe_input_touch_event_raw));
+            touch_point->type = wpe_input_touch_event_type_null;
+        }
+
+        return;
+    }
+    default:
+        g_assert_not_reached ();
+        return;
+    }
+
+    int id = libinput_event_touch_get_seat_slot (touch_event);
+    if (id < 0 || id >= 10)
+        return;
+
+    input_data.last_touch_type = event_type;
+    input_data.last_touch_id = id;
+
+    struct wpe_input_touch_event_raw *touch_point = &input_data.touch_points[id];
+    touch_point->type = event_type;
+    touch_point->time = time;
+    touch_point->id = id;
+
+    if (touch_type == LIBINPUT_EVENT_TOUCH_DOWN
+        || touch_type == LIBINPUT_EVENT_TOUCH_MOTION) {
+        touch_point->x = libinput_event_touch_get_x_transformed (touch_event,
+                                                                 input_data.input_width);
+        touch_point->y = libinput_event_touch_get_y_transformed (touch_event,
+                                                                 input_data.input_height);
+    }
+}
+
+static void
+input_process_events ()
+{
+    g_assert (input_data.libinput);
+    libinput_dispatch (input_data.libinput);
+
+    while (true) {
+        struct libinput_event *event = libinput_get_event (input_data.libinput);
+        if (!event)
+            break;
+
+        enum libinput_event_type event_type = libinput_event_get_type (event);
+        switch (event_type) {
+        case LIBINPUT_EVENT_KEYBOARD_KEY:
+            input_handle_key_event (libinput_event_get_keyboard_event (event));
+            break;
+        case LIBINPUT_EVENT_TOUCH_DOWN:
+        case LIBINPUT_EVENT_TOUCH_UP:
+        case LIBINPUT_EVENT_TOUCH_MOTION:
+        case LIBINPUT_EVENT_TOUCH_FRAME:
+            input_handle_touch_event (event_type,
+                                      libinput_event_get_touch_event (event));
+            break;
+        default:
+            break;
+        }
+
+        libinput_event_destroy (event);
+    }
+}
+
+static int
+input_interface_open_restricted (const char *path, int flags, void *user_data)
+{
+    return open (path, flags);
+}
+
+static void
+input_interface_close_restricted (int fd, void *user_data)
+{
+    close (fd);
+}
+
+static void
+clear_input ()
+{
+    g_clear_pointer (&input_data.libinput, libinput_unref);
+    g_clear_pointer (&input_data.udev, udev_unref);
+}
+
+static gboolean
+init_input ()
+{
+    static struct libinput_interface interface = {
+            input_interface_open_restricted,
```

Comment:
This indentation here looks a bit wonky, please reformat. Consider using designated initializers, too.